batched/system.go

```diff
@@ -0,0 +1,415 @@
+				}
+				if err := s.ds.Sync(lastReprovideKey); err != nil {
+					log.Errorf("could not perform sync of last reprovide time: %v", err)
+				}
+			}
+		}
+	}()
+
+	s.closewg.Add(1)
+	go func() {
+		defer s.closewg.Done()
+
+		var initialReprovideCh, reprovideCh <-chan time.Time
+
+		// If reproviding is enabled (non-zero)
+		if s.reprovideInterval > 0 {
+			reprovideTicker := time.NewTicker(s.reprovideInterval)
+			defer reprovideTicker.Stop()
+			reprovideCh = reprovideTicker.C
+
+			// if there is a non-zero initial reprovide time that was set in the initializer or if the fallback has been
+			if s.initialReprovideDelaySet {
+				initialReprovideTimer := time.NewTimer(s.initalReprovideDelay)
+				defer initialReprovideTimer.Stop()
+
+				initialReprovideCh = initialReprovideTimer.C
+			}
+		}
+
+		for s.ctx.Err() == nil {
+			select {
+			case <-initialReprovideCh:
+			case <-reprovideCh:
+			case <-s.ctx.Done():
+				return
+			}
+
+			err := s.reprovide(s.ctx, false)
+
+			// only log if we've hit an actual error, otherwise just tell the client we're shutting down
+			if s.ctx.Err() == nil && err != nil {
+				log.Errorf("failed to reprovide: %s", err)
+			}
+		}
+	}()
+}
+
+func stopAndEmptyTimer(t *time.Timer) {
+	if !t.Stop() {
+		<-t.C
+	}
+}
+
+func storeTime(t time.Time) []byte {
+	val := []byte(fmt.Sprintf("%d", t.UnixNano()))
+	return val
+}
+
+func parseTime(b []byte) (time.Time, error) {
+	tns, err := strconv.ParseInt(string(b), 10, 64)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return time.Unix(0, tns), nil
+}
+
+func (s *BatchProvidingSystem) Close() error {
+	s.close()
+	err := s.q.Close()
+	s.closewg.Wait()
+	return err
+}
+
+func (s *BatchProvidingSystem) Provide(cid cid.Cid) error {
+	return s.q.Enqueue(cid)
+}
+
+func (s *BatchProvidingSystem) Reprovide(ctx context.Context) error {
+	return s.reprovide(ctx, true)
+}
+
+func (s *BatchProvidingSystem) reprovide(ctx context.Context, force bool) error {
+	if !s.shouldReprovide() && !force {
+		return nil
+	}
+
+	kch, err := s.keyProvider(ctx)
+	if err != nil {
+		return err
+	}
+
+reprovideCidLoop:
+	for {
+		select {
+		case c, ok := <-kch:
+			if !ok {
+				break reprovideCidLoop
+			}
+
+			select {
+			case s.reprovideCh <- c:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+
+	return nil
+}
+
+func (s *BatchProvidingSystem) getLastReprovideTime() (time.Time, error) {
+	val, err := s.ds.Get(lastReprovideKey)
+	if errors.Is(err, datastore.ErrNotFound) {
+		return time.Time{}, nil
+	}
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not get last reprovide time")
+	}
+
+	t, err := parseTime(val)
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not decode last reprovide time, got %q", string(val))
+	}
+
+	return t, nil
+}
+
+func (s *BatchProvidingSystem) shouldReprovide() bool {
+	t, err := s.getLastReprovideTime()
+	if err != nil {
+		log.Debugf(err.Error())
```

Comment:
`log.Debug(err)` or `log.Debugf("foo: %s", err)` or `log.Debugw("getting last reprovide time failed", "error", err)` (preferred).

The current version will:

1. _Always_ try to format an error, even when debug logging is disabled.
2. Treat the error text as a format string (it's not).

Reply:
Went with form 2 here, we can update to go-log v2 in a later PR and switch to Debugw then, it doesn't seem to be present in go-log v1 (unless I missed it)